### ISSUE TITLE
Repository explorer findEntities issue

### ISSUE THREAD

**Issue by planetf1** [odpi/egeria#4526]
Testing 2.6 release, inmem repo, latest repository explorer...
<img width="1464" alt="Screenshot 2021-01-26 at 18 27 52" src="https://user-images.githubusercontent.com/7292002/105887930-3b8a7f00-6004-11eb-970b-55679e1d2742.png">

Selected cocoMDS1 and tried a search for '*'

**Comment by planetf1** [comment]
This error seems to occur regardless of which repo is selected, and with both local and enterprise options.

**Comment by planetf1** [comment]
```
segment1 servers
deserializeUser called with id 6
findByUserId
passport.deserializeUser user is [object Object],err isnull
/servers/* post called /servers/coco/rex/users/pollytasker/instances/entities/by-property-value
Got body: {
  serverName: 'cocoMDS2',
  platformName: 'Core',
  enterpriseOption: true,
  searchText: '*',
  typeName: null,
  classificationNames: []
}
 getServerInfoFromEnv() 1
 getServerInfoFromEnv() 2
Found server name coco
downstream url https://lab-datalake:9443/servers/cocoView1/open-metadata/view-services/rex/users/pollytasker/instances/entities/by-property-value
response.data
{
  class: 'RexSearchResponse',
  relatedHTTPCode: 400,
  exceptionClassName: 'org.odpi.openmetadata.viewservices.rex.api.ffdc.RexViewServiceException',
  actionDescription: 'findEntities',
  exceptionErrorMessage: 'The repository explorer view service operation findEntities encountered a problem connecting to the repository, the message is OMRS-REPOSITORY-400-040 An invalid property has been found in a metadata instance stored in repository cocoMDS1 during the findEntitiesByPropertyValue operation',
  exceptionErrorMessageId: 'OMVS-REPOSITORY-EXPLORER-400-007',
  exceptionErrorMessageParameters: [
    'findEntities',
    'OMRS-REPOSITORY-400-040 An invalid property has been found in a metadata instance stored in repository cocoMDS1 during the findEntitiesByPropertyValue operation'
  ],
  exceptionSystemAction: 'The system could not complete a repository operation due to the error indicated.',
  exceptionUserAction: 'Look at the error message and check the configuration and state of the repository server.'
}
```

**Comment by planetf1** [comment]
Note I don't rememeber logging on as polly tasker - however I re-logged on and get the same result with garygeeke

Another issue is that cocoMDS5 is failing to work with REX. It is defined as a repository proxy, with an inmem setup, is this valid? If so it should work, if not we should remove from the rex configuration list?

<img width="1467" alt="Screenshot 2021-01-26 at 18 41 07" src="https://user-images.githubusercontent.com/7292002/105889543-30d0e980-6006-11eb-961a-5efd5ebe3016.png">

**Comment by planetf1** [comment]
cc: @grahamwallis

**Comment by grahamwallis** [comment]
One comment about the first test - "*" is not a valid regex so I think the RepositoryContentValidator is probably catching an exception from the matches() method. Most probably it is catching:

java.util.regex.PatternSyntaxException: Dangling meta character '*' near index 0

However, this test does show the difficulty of pinning down the exact cause of the problem - i.e. the exception being reported by Rex indicates an invalid property - which (if I am right about the cause above) is a bit misleading in itself. Digging through the code I would guess that cocoMDS2 is running an in-memory repository and contains an entity that is being inspected as part of findEntitiesByPropertyValue. The matching of properties against the searchCriteria is done by verifyInstancePropertiesMatchSearchCriteria and if this gets any exception from any of the underlying switch cases it will throw a '400-040'. Beyond that it is impossible to precisely pin down the cause of the problem - the try catch is fairly generic and doesn't have access to the propertyName or entity typeName. The RepositoryContentValidator method could be refactored to enable the property category and propertyName to be included in the exception and to record more precisely the nature of the problem. The exception could also be caught in the calling method, InMemoryOMRSMetadataCollection::findEntitiesByPropertyValue, and augmented with the type name and re-thrown.

**Comment by grahamwallis** [comment]
Regarding the second point (RepositoryProxy) - is this related to the problem I worked on yesterday about the type system not being complete and well-formed? If so I think the changes I put into Tex in PR #4519 would help - they at least would help to isolate the cause of the problem.

I only included those changes in Tex and it occurred to me last night that Rex and Dino will need similar FFDC so I will add them today.

**Comment by planetf1** [comment]
I'll test with some other searches.

I'd not noticed either issue in previous releases, but I would need to go back and retest to be sure. I may have not noticed.

* If the search string isn't a valid regex we probably should sanitize in the backend API call. Otherwise we may be exposed to injection attacks if other invalid syntax is passed?
* Agree on error handling
* The repository proxy angle is overlapping -- though here this was just with coco pharma, as opposed to another developer's experiments with the atlas connector. However I not noticed this issue before

**Comment by grahamwallis** [comment]
It's probably much harder than it sounds to validate or sanitize the regex. About the most that could be done I guess would be to perform a 'matches()' against a dummy string and see whether the parser blows its top.

**Comment by grahamwallis** [comment]
The NPE issue is addressed by #4544

**Comment by planetf1** [comment]
Retesting with the latest 4.6 code (which includes your commit on sparse exceptions) and now seeing the original issue..

This of course is repeating the invalid search term

<img width="1259" alt="Screenshot 2021-02-01 at 18 12 52" src="https://user-images.githubusercontent.com/7292002/106499959-2bb5e380-64b9-11eb-9dd5-76b565c9f59a.png">

Leaving this open. will test some more & check again with master .

**Comment by grahamwallis** [comment]
To me it looks like the regex exception is being mapped into something too generic - that message appears to suggest an invalid property in an instance whereas the real problem is an invalid parameter and we should at least indicate which parameter (which needs a view-service level mapping to what the user actually sees on the screen, rather than the API parameter name which would be meaningless to the user) and ideally it would be nice to see (at least some of) the original exception message about the regex.

**Comment by planetf1** [comment]
I suggest we don't need to fix this error for 2.6 - I'll add a note to the readme about regex, but it would be a useful usability improvement

**Comment by planetf1** [comment]
I repeated this test, albeit on a simpler environment, searching for '*' when I have instances loaded, and get:
```
Could not find entities (status : 400). The repository explorer view service operation findEntities encountered a problem connecting to the repository, the message is OMRS-REPOSITORY-400-055 An unexpected java.util.regex.PatternSyntaxException exception was received from a repository connector during the findEntitiesByPropertyValue operation which had message: Dangling meta character '*' near index 0
*
^

System detail: The system could not complete a repository operation due to the error indicated.

Recommended user action: Look at the error message and check the configuration and state of the repository server.
```

Searching for `.*` as a valid regex pattern works ok

As such I will close for now
